framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -1822,19 +1823,95 @@ protected void commonAssignmentCheck(
+            String varTypeString;
+            if (shouldPrintVerbose(varType, valueType)) {
                 valueTypeString = valueType.toString(true);
                 varTypeString = varType.toString(true);
+            } else {
+                valueTypeString = valueType.toString();
+                varTypeString = varType.toString();
             }
             checker.report(Result.failure(errorKey, valueTypeString, varTypeString), valueTree);
         }
     }
 
+    /**
+     * Return whether or not the verbose toString should be used when printing the two annotated
+     * types.
+     *
+     * <p>If any simple toString of any annotated type in the first atm or second atm is equal to
+     * any toString of any annotated type in the first or second atm and if the verbose toStrings
+     * differ, then the verbose toString should be used.
+     *
+     * @param atm1 the first atm
+     * @param atm2 the second atm
+     * @return whether or not the verbose toString should be used when printing the two annotated
+     *     types.
+     */
+    private boolean shouldPrintVerbose(AnnotatedTypeMirror atm1, AnnotatedTypeMirror atm2) {
+        String atm1ToString = atm1.toString();
+        String atm2ToString = atm2.toString();
+        // If both types as strings are the same, use verbose toString.
+        if (atm2ToString.equals(atm1ToString)
+                // or if neither string contains an annotation
+                || (!atm2ToString.contains("@") && !atm1ToString.contains("@"))) {
+            return true;
+        }
+
+        SimpleAnnotatedTypeScanner<Boolean, Void> checkForMismatchedToStrings =
+                new SimpleAnnotatedTypeScanner<Boolean, Void>() {
+                    Map<String, List<AnnotatedTypeMirror>> map = new HashMap<>();
+
+                    @Override
+                    protected Boolean reduce(Boolean r1, Boolean r2) {
+                        if (r1 == null) {
+                            if (r2 == null) {
+                                return false;
+                            } else {
+                                return r2;
+                            }
+                        } else if (r2 == null) {
+                            return r1;
+                        } else {
+                            return r1 || r2;
+                        }
+                    }
+
+                    @Override
+                    protected Boolean defaultAction(AnnotatedTypeMirror type, Void avoid) {
+                        if (type == null) {
+                            return false;
+                        }
+                        String simple = type.toString();
+                        List<AnnotatedTypeMirror> list = map.get(simple);
+                        if (list == null) {
+                            list = new ArrayList<>();
+                            map.put(simple, list);
+                        }
+                        String verbose = type.toString(true);
```

Comment:
Please comment that this isn't List.contains because it's testing toString().equals() instead of equals().